### PR TITLE
Enable GC loading from a shared library

### PR DESCRIPTION
**Pull request by eightbitraptor**
This PR replaces https://github.com/ruby/ruby/pull/10302 as a method for dynamically loading a shared GC library. This new approach has been guided by feedback from @kateinoigakukun and @nobu 

The approach taken here is as follows:

- Enable this feature by configuring with `--with-shared-gc`
- When enabled, Ruby will check for the presence of the environment variable `RUBY_GC_LIBRARY_PATH`
- If that path exists and points to a valid shared object
  - Open the shared object (using `dln_open`), and map the `Init_GC` function to a function map
  - call `Init_GC` as part of `rb_objspace_alloc` when initialising the GC
- If that path does not point to a valid shared object
  - Degrade gracefully to initialising the existing Ruby objspace and GC
  
When Ruby is configured without the `--with-shared-gc` flag, no extra code is compiled and so existing behaviour is maintained without any performance penalty.

Enabling this feature, but not populating `RUBY_GC_LIBRARY_PATH` will incur a small performance penalty due to the overhead of using `getenv` to check for the environment variable.

The use of `dln.c` rather than using `dlopen` directly, should enable this feature to be supported on all platforms that support loading shared libraries, rather than the initial implementation, which explicitly only supported Linux and MacOS.

### CODE REVIEW

File: gc.c [hunk 1]
```diff
@@ -1884,6 +1884,52 @@ rb_gc_initial_stress_set(VALUE flag)
...
+    char *gc_so_path = getenv("RUBY_GC_LIBRARY_PATH");
+    if (!gc_so_path) {
+        map->init = Alloc_GC_impl;
+        return;
+    }
+
+    void *h = dln_open(gc_so_path);
+    if (!h) {
+        rb_bug(
+            "ruby_external_gc_init: Shared library %s cannot be opened.\n",
```
Comment (nobu) [inline gc.c:1908]: `rb_bug` appends a newline.
```suggestion
            "ruby_external_gc_init: Shared library %s cannot be opened.",
```

File: gc.c [hunk 2]
```diff
@@ -1884,6 +1884,52 @@ rb_gc_initial_stress_set(VALUE flag)
...
+            gc_so_path
+        );
+    }
+
+    void *gc_init_func = dln_symbol(h, "Init_GC");
+    if (!gc_init_func) {
+        rb_bug("ruby_external_gc_init: Init_GC func not exported by library");
```
Comment (nobu) [inline gc.c:1915]: The failing shared library path is not needed?

```suggestion
        rb_bug("ruby_external_gc_init: Init_GC func not exported by library %s.",
            gc_so_path);
```

File: dln.h [hunk 3]
```diff
@@ -25,6 +25,7 @@ RUBY_SYMBOL_EXPORT_BEGIN
 char *dln_find_exe_r(const char*,const char*,char*,size_t DLN_FIND_EXTRA_ARG_DECL);
 char *dln_find_file_r(const char*,const char*,char*,size_t DLN_FIND_EXTRA_ARG_DECL);
 void *dln_load(const char*);
+void * dln_open(const char *file);
```
Comment (nobu) [inline dln.h:28]:
```suggestion
void *dln_open(const char *file);
```

File: configure.ac [hunk 4]
```diff
@@ -4657,6 +4659,7 @@ config_summary "compiler"            "$CC"
 config_summary "with thread"         "$THREAD_MODEL"
 config_summary "with coroutine"      "$coroutine_type"
 config_summary "enable shared libs"  "$ENABLE_SHARED"
+config_summary "enable shared GC"    "$gc_support"
```
Comment (nobu) [inline configure.ac:4662]: Why not `$with_shared_gc`?

**Review by kateinoigakukun — APPROVED** [review]

File: tool/m4/ruby_shared_gc.m4 [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+AS_IF([test "$with_shared_gc" = "yes"], [
+    AC_MSG_RESULT([yes])
+    gc_support="yes"
+
+    AC_DEFINE([USE_SHARED_GC], [1])
+], [
+    AC_MSG_RESULT([no])
+    gc_support="no"
+
+    AC_DEFINE([USE_SHARED_GC], [0])
```
Comment (nobu) [inline tool/m4/ruby_shared_gc.m4:20]: This `gc_support` word feels like GC is not opted out if it is "no".

```suggestion
    AC_MSG_RESULT([yes])
    AC_DEFINE([USE_SHARED_GC], [1])
], [
    AC_MSG_RESULT([no])
    with_shared_gc="no"
    AC_DEFINE([USE_SHARED_GC], [0])
```

File: configure.ac [hunk 6]
```diff
@@ -4656,6 +4658,7 @@ config_summary "target OS"           "$target_os"
 config_summary "compiler"            "$CC"
 config_summary "with thread"         "$THREAD_MODEL"
 config_summary "with coroutine"      "$coroutine_type"
+config_summary "with shared GC"      "$gc_support"
```
Comment (nobu) [inline configure.ac:4661]:
```suggestion
config_summary "with shared GC"      "$with_shared_gc"
```

**Review by nobu — CHANGES_REQUESTED** [review]

**Pull request merged**